altair/vegalite/v5/schema/channels.py

```diff
@@ -179,7 +179,7 @@ def to_dict(
                 # We still parse it out of the shorthand, but drop it here.
                 parsed.pop("type", None)
             elif not (type_in_shorthand or type_defined_explicitly):
-                if _is_pandas_dataframe(context.get("data", None)):
+                if isinstance(context.get("data", None), nw.DataFrame):
```

Comment:
with the changes since #3452, by the time the code gets here, `context.get('data')` would have been "Narwhalified", and not a raw pandas dataframe anymore

Reply:
thanks @dangotbanned !

agree on the commit you added

I've factored out `context.get('data', None)`, I'm just not sure about the annotation, as this can be a lot more than `SomeFrame` - check the return type of `_prepare_data`

Reply:
Thanks @MarcoGorelli and for pointing to `_prepare_data`, that really helped.

So if I've understood this correctly, starting in `Chart`:

https://github.com/vega/altair/blob/0062e6228134f176bb9fbaa944724039bb6b043a/altair/vegalite/v5/api.py#L3661-L3663

`context["data"]` is set in `TopLevelMixin.to_dict`:

https://github.com/vega/altair/blob/0062e6228134f176bb9fbaa944724039bb6b043a/altair/vegalite/v5/api.py#L1792-L1799

Which tries with `narwhals`:
https://github.com/vega/altair/blob/0062e6228134f176bb9fbaa944724039bb6b043a/altair/utils/core.py#L486

And prepares in:

https://github.com/vega/altair/blob/0062e6228134f176bb9fbaa944724039bb6b043a/altair/vegalite/v5/api.py#L253-L255

@MarcoGorelli 
> I'm just not sure about the annotation, as this can be a lot more than `SomeFrame` - check the return type of `_prepare_data`

So with this in mind - I guess this is more of a question now - can we assume the type of `context["data"]` has been **narrowed at all** by this point?

I would have thought for the cases `narwhals` can handle, they'd at least be wrapped by now.
If so, that would be information I'd find useful if it can be conveyed through typing.

Please let me know if I'm missing something here

Reply:
> I would have thought for the cases narwhals can handle, they'd at least be wrapped by now

Yup! 👍 

> If so, that would be information I'd find useful if it can be conveyed through typing.

Would `: nw.DataFrame | Any` be acceptable to you?

Reply:
> Would `: nw.DataFrame | Any` be acceptable to you? For someone reading this particular function, that might be all they need to know?

Yeah @MarcoGorelli that seems the simplest